include/Basic/SourceManager.hpp

```diff
@@ -0,0 +1,273 @@
+        if (_buffer)
+            return _buffer->getBuffer();
+        return std::nullopt;
+    }
+
+    /// Set the MemoryBuffer for this ContentCache.
+    void setBuffer(std::unique_ptr<llvm::MemoryBuffer> buffer)
+    {
+        _isBufferInvalid = false;
+        _buffer = std::move(buffer);
+    }
+
+    void setFileName(llvm::StringRef fileName)
+    {
+        _fileName = fileName;
+    }
+
+    llvm::StringRef getFileName() const
+    {
+        return _fileName;
+    }
+};
+
+///
+/// @class FileInfo
+/// @brief Contains concrete information about a file.
+///
+/// The FileInfo class contains information about a file that has been loaded by
+/// the SourceManager. It contains the SourceLocation of the import directive
+/// that brought in the file, and a reference to the ContentCache object that
+/// contains the content of the file.
+///
+class FileInfo {
+
+private:
+    /// The location of the import directive that brought in the file.
+    /// If the FileInfo is for the main file this location is invalid (ID == 0).
+    SourceLocation _importLoc;
+
+    /// The content cache of the file.
+    ContentCache const *_contentCache;
+
+public:
+    FileInfo()
+        : _importLoc(0)
+        , _contentCache(nullptr) {};
+
+    /// Create a FileInfo object with the given SourceLocation and ContentCache.
+    static FileInfo
+    get(SourceLocation includeLoc, ContentCache &contentCache,
+        llvm::StringRef fileName)
+    {
+        FileInfo info;
+
+        info._importLoc = includeLoc;
+        info._contentCache = &contentCache;
+        contentCache.setFileName(fileName);
+        return info;
+    }
+
+    SourceLocation getImportLoc() const
+    {
+        return _importLoc;
+    }
+
+    ContentCache const &getContentCache() const
+    {
+        return *_contentCache;
+    }
+
+    llvm::StringRef getFileName() const
+    {
+        return _contentCache->getFileName();
+    }
+};
+
+///
+/// @class SourceLocEntry
+/// @brief Represents a specific location in the source code.
+///
+/// A SourceLocEntry is an object that represents a specific location
+/// in the source code. It is used to refer to a specific character in a
+/// specific file. The SourceManager class is responsible for creating and
+/// interpreting SourceLocEntry objects.
+///
+/// A SourceLocEntry is basicly an offset into the complete source code. The
+/// SourceManager knowings to which file this offset belongs can interpret it
+/// and provide useful informations from it.
+///
+class SourceLocEntry {
+
+private:
+    /// The offset of the source location.
+    uint32_t _offset;
+
+    /// The file information for the source location.
+    FileInfo _file;
+
+public:
+    SourceLocEntry(uint32_t offset, FileInfo const &file)
+        : _offset(offset)
+        , _file(file)
+    {
+    }
+
+    /// Returnn a const reference to the FileInfo object.
+    FileInfo const &getFileInfo() const
+    {
+        return const_cast<SourceLocEntry *>(this)->getFile();
```

Comment:
```suggestion
        return _file;
```
is that not easier??